src/node/blockstorage.cpp

```diff
@@ -611,7 +611,12 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
 
     m_blockfile_info[nFile].AddBlock(nHeight, nTime);
     if (fKnown) {
-        m_blockfile_info[nFile].nSize = std::max(pos.nPos + nAddSize, m_blockfile_info[nFile].nSize);
+        // pos.nPos points at the offset of the block data in the blk file. that already accounts for
+        // the serialization header present in the file (the 4 magic message start bytes + the 4 length bytes = 8 bytes).
+        // nAddSize already has 8 added to it too. we subtract 8 to avoid over accounting for the serialization
+        // header. such over accounting was a long standing bug that added undesirable 8 byte gaps into blk data files following
+        // a -reindex operation. for more info, see https://github.com/bitcoin/bitcoin/issues/21379
+        m_blockfile_info[nFile].nSize = std::max(pos.nPos + nAddSize - 8, m_blockfile_info[nFile].nSize);
```

Comment:
Thanks for the review!

I resonate with what you said about not subtracting 8 within `BlockManager::FindBlockPos` for code clarity. Subtracting 8 was just the first simplest fix that I thought of.

Building off of your idea, instead of making two different calls, we could get the same effect by doing something like the following in `BlockManager::SaveBlockToDisk`:

```
    const auto position_known = dbp != nullptr;
    const auto additional_size = nBlockSize + (position_known ? 0 : 8);
    if (!FindBlockPos(blockPos, additional_size, nHeight, active_chain, block.GetBlockTime(), position_known)) {
```

Reply:
made updates for this in 66b8d3c23d999651c79ac03e7558d4248a200ba1